modules/broker-core/src/main/java/org/wso2/broker/core/MessagingEngine.java

```diff
@@ -168,22 +155,31 @@ void publish(Message message) throws BrokerException {
                     LOGGER.info("Dropping message since no queues found for routing key " + routingKey);
                     message.release();
                 } else {
-                    boolean published = false;
-                    for (Binding binding : bindingSet.getUnfilteredBindings()) {
-                        published |= pushToInMemoryQueue(message, binding);
-                    }
+                    try {
+                        sharedMessageStore.add(message);
+                        Set<String> uniqueQueues = new HashSet<>();
+                        for (Binding binding : bindingSet.getUnfilteredBindings()) {
+                            uniqueQueues.add(binding.getQueue().getName());
+                        }
 
-                    for (Binding binding : bindingSet.getFilteredBindings()) {
-                        if (binding.getFilterExpression().evaluate(metadata)) {
-                            published |= pushToInMemoryQueue(message, binding);
+                        for (Binding binding : bindingSet.getFilteredBindings()) {
+                            if (binding.getFilterExpression().evaluate(metadata)) {
+                                uniqueQueues.add(binding.getQueue().getName());
+                            }
                         }
+                        // Unique queues can be empty due unmatching selectors.
+                        if (!uniqueQueues.isEmpty()) {
+                            boolean published = publishToQueues(message, uniqueQueues);
```

Comment:
Do we need a local variable for this? Can't directly use it on if?

Reply:
Updated the code segment.